src/helics/core/ActionMessage.cpp

```diff
@@ -281,12 +283,105 @@ int ActionMessage::serializedByteCount() const
+    }
     return data;
 }
 
+static bool isBinaryData(std::string_view str)
+{
+    return std::any_of(str.begin(), str.end(), [](const auto& c) {
+        return ((c < 32) || (c == 34) || (c > 126));
+    });
+}
+
+static bool isEscapableData(std::string_view str)
+{
+    return std::all_of(str.begin(), str.end(), [](const auto& c) {
+        return ((c >= 32 && c <= 126) || (c == '\t') || (c == '\n'));
+    });
+}
+
+/** encode the string in base64 if needed otherwise just return the string*/
+static std::string b64Encode(std::string_view str2encode)
+{
+
+    if (isBinaryData(str2encode)) {
+        if (isEscapableData(str2encode)) {
+            return std::string(str2encode);
+        } else {
+            return std::string("b64[") +
+                gmlc::utilities::base64_encode(reinterpret_cast<const unsigned char*>(
+                                                   str2encode.data()),
+                                               str2encode.size()) +
+                ']';
+        }
+
+    } else {
+        return std::string(str2encode);
+    }
+}
+
+// check if the string has a base 64 encoding wrapper
+static int hasB64Wrapper(std::string_view str)
+{
+    if (str.compare(0, 4, "b64[") == 0) {
+        return 4;
+    }
+    return 0;
+}
+
+// decode a potentially base 64 encoded strings
+static std::string b64Decode(std::string_view stringToDecode)
+{
+    if (stringToDecode.empty()) {
+        return std::string{};
+    }
+    auto offset = hasB64Wrapper(stringToDecode);
+    if (offset != 0) {
+        stringToDecode.remove_suffix(1);
+        return gmlc::utilities::base64_decode_to_string(std::string(stringToDecode), offset);
+    }
+
+    // move is required since you are returning the rvalue and we want to move from the rvalue input
+    return std::string(stringToDecode);
+}
+
+std::string ActionMessage::to_json_string() const {
+    Json::Value packet;
+    packet["version"] =
+        HELICS_VERSION_MAJOR * 10000 + HELICS_VERSION_MINOR * 100 + HELICS_VERSION_PATCH;
+    packet["command"] = static_cast<int>(messageAction);
+    packet["messageId"] = messageID;
+    packet["sourceId"] = source_id.baseValue();
+    packet["sourceHandle"] = source_handle.baseValue();
+    packet["destId"] = dest_id.baseValue();
+    packet["destHandle"] = dest_handle.baseValue();
+    packet["counter"] = counter;
+    packet["flags"] = flags;
+    packet["sequenceId"] = sequenceID;
+    packet["actionTime"] = actionTime.getBaseTimeCode();
+    if (messageAction == CMD_TIME_REQUEST) {
+        packet["Te"] = Te.getBaseTimeCode();
+        packet["Tdemin"] = Tdemin.getBaseTimeCode();
+        packet["Tso"] = Tso.getBaseTimeCode();
+    }
+    packet["payload"] = b64Encode(payload.to_string());
+    packet["stringCount"] = stringData.size();
+    if (!stringData.empty()) {
+        packet["strings"] = Json::arrayValue;
+        for (const auto& str : stringData){
+            packet["strings"].append(b64Encode(str));
```

Comment:
What do you think about skipping the base64 encoding and let jsoncpp do its escaping as needed for string values? I think either way jsoncpp will be checking strings to see if it needs to escape them and creating a new std::string object, though we'd reduce the number of string objects and checks on our side.

I guess the downside would be encoded size, with the worst case for not encoding as base64 being a large blob of data that only has byte values below 32; and the worst case for encoding as base64 would be if there is a large blob of binary data with only a few characters below 32 triggering the base64 encoding resulting in the (~33%?) overhead of base64 encoding.

Reply:
I tried this seems to work.  Probably will be slower in a few cases but might not matter that much

Reply:
The binary strings tests takes quite a bit longer but that might be expected being that is 50000 bytes which would all need to be escaped